pkg/etcd/etcd.go

```diff
@@ -547,30 +584,41 @@ func snapshotDir(config *config.Control) (string, error) {
 	return config.EtcdSnapshotDir, nil
 }
 
+// snapshot attempts to save a new snapshot to the configured directory, and then clean up any old
+// snapshots in excess of the retention limits.
 func (e *ETCD) snapshot(ctx context.Context) {
-	snapshotTime := time.Now()
-	logrus.Infof("Snapshot retention check")
+	status, err := e.client.Status(ctx, endpoint)
+	if err != nil {
+		logrus.Errorf("Failed to check etcd status for snapshot: %v", err)
+		return
+	}
+
+	if status.IsLearner {
+		logrus.Warnf("Skipping snapshot: not supported for learner")
+		return
+	}
+
 	snapshotDir, err := snapshotDir(e.config)
 	if err != nil {
 		logrus.Errorf("Failed to get the snapshot dir: %v", err)
 		return
 	}
-	logrus.Infof("Taking etcd snapshot at %s", snapshotTime.String())
 	sManager := snapshot.NewV3(nil)
 	tlsConfig, err := toTLSConfig(e.runtime)
 	if err != nil {
-		logrus.Errorf("Failed to get tls config for etcd: %v", err)
+		logrus.Errorf("Failed to get tls config for etcd snapshot: %v", err)
 		return
 	}
 	etcdConfig := etcd.Config{
```

Comment:
I think we also need to keep the etcd client in check by checking and setting default values for the timeouts:

```
	// DialTimeout is the timeout for failing to establish a connection.
	DialTimeout time.Duration `json:"dial-timeout"`

	// DialKeepAliveTime is the time after which client pings the server to see if
	// transport is alive.
	DialKeepAliveTime time.Duration `json:"dial-keep-alive-time"`

	// DialKeepAliveTimeout is the time that the client waits for a response for the
	// keep-alive probe. If the response is not received in this time, the connection is closed.
	DialKeepAliveTimeout time.Duration `json:"dial-keep-alive-timeout"`
```